### ISSUE TITLE
BPM Not changing

### ISSUE THREAD

**Issue by TomasMax22** [Fluto/TaikoMods#16]
Im using the latest version in mono
While playing rokuchounen to ichiya monogatari almost in the end the song slows a little bit but the BPM does not and it gets desynchronizated till the end of the song
[Rokuchounen to Ichiya Monogatari.zip](https://github.com/Fluto/TakoTako/files/9389730/Rokuchounen.to.Ichiya.Monogatari.zip)
.

**Comment by alexankitty** [comment]
I took some time to look into this. I found that the song starts off 200ms off, and then of course as you witnessed at the end, the BPM change kinda just doesn't happen and then the whole song is off by several beats.
This appears to be a problem specifically with tja2bin which the project leverages to convert them into fumen format, and still occurred even when I converted directly instead of through tjaconvert. Also note that using a tjaplayer it's perfect all the way through.. My worry with this is that it's also potentially messing with a lot of other songs.

Though the only way to really fix is going to be reverse engineering how the fumen format is built. And it seems the internet is very scant with information regarding it. The most I know is that they're in gzip format, and it seems like each measure is padded by a block of FF bytes. Though according to other projects like Okku which is used for converting to fumen for 3DS, the format has a few limitations.

**Comment by joshuacwnewton** [comment]
> My worry with this is that it's also potentially messing with a lot of other songs.

I can confirm that this is the case. 

I recently tried to convert some TJAs of the recently released (~2022) AC-only songs, and every song with BPM changes resulted in desynchronization. 

Some very clear examples of nearly unplayable songs:

- Kyourandotou
- R.I.P. Hero
- Dragon of Time ~Kokuryu~
- Zouo to Shuuaku no Hanataba
- Aragami
- Roppon no Bara to Sai no Uta

**Comment by joshuacwnewton** [comment]
> Though according to other projects like [Okku](https://www.teampapafox.org/projects/okku3ds.html) which is used for converting to fumen for 3DS, the format has [a few limitations](https://www.teampapafox.org/documents.html).

Okku's [TJA-Fumen conversion document](https://www.teampapafox.org/documents.html) does seem like a nice resource. So, for kicks, I tried messing around with Okku's fumen outputs (decompressing LZ11 using [`nzlss`](https://github.com/magical/nlzss), then recompressing in gzip) but those experiments went nowhere, hehe.

----

Separately, some cursory searching led to a few other resources such as [`fumen2osu.py`](https://github.com/KatieFrogs/fumen-tools/blob/main/fumen2osu/fumen2osu.py), [`replace_timing.py`](https://github.com/cainan-c/TaikoFumenTimingReplace/blob/main/replace_timing.py), etc. 

Out of curiosity, I tried running `fumen2osu.py` on "proper" console/AC fumens (that correctly handle BPM changes), and the script was able to parse TDMX `.bin` files just fine. For example:

![image](https://user-images.githubusercontent.com/16181459/230799035-100ff2ce-07b4-44a0-9f38-06cdf732a246.png)

The source code (and the structure of the `song` object) have a lot of valuable details to them. It makes me want to learn more about how exactly the script parses fumen files. 

(At the very least, if there are Python scripts to parse existing variable-BPM TDMX fumens, then perhaps that knowledge can be applied in reverse to _create_ variable-BPM fumen files, too.)

**Comment by joshuacwnewton** [comment]
> And it seems the internet is very scant with information regarding it.

I'm going to try to deeply document the `readFumen()` function, that way we hopefully have a good resource on how fumen binaries are constructed, and how they can be parsed. 

I'll be using the [`tja2fumen.py`](https://github.com/joshuacwnewton/fumen-tools/blob/jn/develop-tja2fumen.py/tja2fumen/tja2fumen.py) file I've created in my fork of the `fumen-tools` repo for documentation/experimentation.

----

<details>
<summary> Experiment 1: Comparing TJA2BIN output to official Nijiro conversion </summary>

----

> * [ ] Try converting a TJA to `.bin`, then comparing that `.bin` to an "official" console/AC fumen `.bin` of the same song. (By loading them both into the `fumen2osu.py` script as `song` objects, then seeing where things are different.) This will help to understand where `tja2bin` is going wrong when it comes to BPM changes.

I tried comparing files using Rokuchounen to Ichiya Monogatari as an example: 

- `210239496_m.bin` (tja2bin conversion)
- `ia6cho_m.bin` (Nijiro conversion)
 
But we immediately run into a quirk: the script's `readFumen()` function couldn't parse the `210239496_m.bin` at all. The metadata it parses is complete nonsense:

```
Total measures: 3103431119, has branches, Little-endian
Measure #1 (normal) at 230-d8268 (bpm: -4.922007519780891e-06, fumenOffset: 1.1650809679836072e-28, offset: -48760591899.84413, gogo: 227, hidden: 64, speed: 29782638592)
Total notes: 36866
Note #1 at 238-24f
Traceback (most recent call last):
  File "C:/Program Files/JetBrains/PyCharm Community Edition 2022.3/plugins/python-ce/helpers/pydev/pydevd.py", line 1496, in _exec
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "C:\Program Files\JetBrains\PyCharm Community Edition 2022.3\plugins\python-ce\helpers\pydev\_pydev_imps\_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "C:\Users\joshu\repos\fumen-tools\tja2fumen\tja2fumen.py", line 254, in <module>
    parsedSong = readFumen(inputFile, arguments.order, arguments.debug)
  File "C:\Users\joshu\repos\fumen-tools\tja2fumen\tja2fumen.py", line 119, in readFumen
    raise ValueError("Error: Unknown note type '{0}' at offset {1}".format(
ValueError: Error: Unknown note type '379B9E75' at offset 0x238
```

The Nijiro converted file can be parsed just fine, though:

```
Total measures: 116, no branches, Little-endian

Measure #1 at 230-238 (bpm: 82.0007095336914, fumenOffset: 885.1959838867188, offset: 3811.999926984209, gogo: False, hidden: False, speed: 1)
Total notes: 0

Measure #2 at 270-278 (bpm: 73.08783721923828, fumenOffset: 1259.98095703125, offset: -65.54633406730372, gogo: False, hidden: False, speed: 1)
Total notes: 0
[...]
```

So, already, tja2bin is doing something... different. To the point where I don't know if I can rely on it for debugging purposes.

</details>

----

I'll keep editing this comment with my experimentation as to not dominate this issue with spam/notifications.

**Comment by joshuacwnewton** [comment]
I'm maybe 80% of the way there developing a new tja2bin converter in Python ([`tja2fumen`](https://github.com/joshuacwnewton/fumen-tools/tree/jn/develop-tja2fumen.py/tja2fumen)) Right now, I can:

- [parse `.tja` files](https://github.com/joshuacwnewton/fumen-tools/blob/9e2d40315760eb92a363268b6a6121ed26fdcffa/tja2fumen/parsers.py#L18-L51) (90% done)
- [Convert the TJA chart data info to the fumen format](https://github.com/joshuacwnewton/fumen-tools/blob/9e2d40315760eb92a363268b6a6121ed26fdcffa/tja2fumen/converters.py#L96-L156) (60% done)
- [Write the info to `.bin` files ](https://github.com/joshuacwnewton/fumen-tools/blob/9e2d40315760eb92a363268b6a6121ed26fdcffa/tja2fumen/writers.py#L5-L63)(100% done).

I've been using Rokuchounen to Ichiya Monogatari as a test song. But, I'm still very confused by the BPM changes near the end.

<details>

----

So, the song stays at a constant ~186 BPM for the most part. Then, we get to this section:

![image](https://user-images.githubusercontent.com/16181459/233501242-4eaa81e5-9697-4f7e-af59-ac251632c7eb.png)

[Wikiwiki.jp/taiko-fumen](https://wikiwiki.jp/taiko-fumen/%E5%8F%8E%E9%8C%B2%E6%9B%B2/%E3%81%8A%E3%81%AB/%E5%85%AD%E5%85%86%E5%B9%B4%E3%81%A8%E4%B8%80%E5%A4%9C%E7%89%A9%E8%AA%9E) says the BPM changes are: 186, 178, 168, 159, 144, 126, 107, 60.61, 186

When we parse an official fumen file, we get: 185.997, 178.001, 167.999, 159.004, 143.999. 126.000, 106.999, 60.607, 185.999. (Basically the exact same BPMs!)

![image](https://user-images.githubusercontent.com/16181459/233501765-a9c2581c-0785-43fc-b69f-2837c52947e0.png)

However, the TJA file uses the BPMs 186, 160, 142, 125, 115, and 59 instead:

<details>
<summary> TJA chart info </summary>

```bash
2020101020002222,
101120101011
#BPMCHANGE 160
#SCROLL 1.1625
2020,
#MEASURE 5/4
#SCROLL 1
11
#BPMCHANGE 142
21
#BPMCHANGE 125
33

#GOGOEND
#BPMCHANGE 115
33
#BPMCHANGE 59
00,

#GOGOSTART
#MEASURE 2/4
#BPMCHANGE 186
20112010,
#MEASURE 4/4
2021202120212021,
2020101020001122,
```

</details>

![image](https://user-images.githubusercontent.com/16181459/233503861-6490edb7-982b-4c94-add2-1738f258326b.png)

![image](https://user-images.githubusercontent.com/16181459/233503901-e6e28555-8113-43c5-ace7-45d934fae4d8.png)



The BPMs are completely different! They don't line up at all with "official" fumens.

Yet, somehow, when this TJA file is loaded into a TJAPlayer, it plays just fine with no desynchronization. Why??? What is the TJA player doing to smooth out these BPM changes? How does it just work???

----

</details>

I could easily just edit the TJA file to use the proper fumen BPMs from wikwiki, and the converter would probably work perfectly. But, that's not my main goal here. I'll keep working to understand the differences between the "official" fumen BPMs and the TJA BPMs.

**Comment by joshuacwnewton** [comment]
I understand the problem, and discovered that someone has already created a solution:

- The issue comes from songs that have `#BPMCHANGE` commands _in the middle of a measure_.
- To solve this problem, a user on the Taiko Modding Discord server [wrote a Python script](https://discord.com/channels/859872100860035113/859879261217095691/986064238369964082) that wraps Pulsar's `tja2bin` converter (the same one used by TakoTako).
- This script modifies the `.tja` file, and inserts `#MEASURE`, `#BARLINEOFF` and `#BARLINEON` commands near the `#BPMCHANGE` commands. 
- Somehow, this fixes the BPM issues.

For Rokuchounen to Ichiya Monogatari, the changes are like this:

<details>
<summary> Before: </summary>

```bash
2021202120212021,
2020101020002222,
101120101011
#BPMCHANGE 160
#SCROLL 1.1625
2020,
#MEASURE 5/4
#SCROLL 1
11
#BPMCHANGE 142
21
#BPMCHANGE 125
33

#GOGOEND
#BPMCHANGE 115
33
#BPMCHANGE 59
00,

#GOGOSTART
#MEASURE 2/4
#BPMCHANGE 186
20112010,
#MEASURE 4/4
2021202120212021,
2020101020001122,
```
</details>

<details>
<summary> After: </summary>

```bash
2021202120212021,
2020101020002222,
#MEASURE 2/4
10112010,
#BARLINEOFF
#BPMCHANGE 178
#SCROLL 1.04
#MEASURE 1/4
1011,
#BPMCHANGE 168
#SCROLL 1.11
#MEASURE 1/4
2020,
#BARLINEON
#MEASURE 4/4
#MEASURE 5/4
#BPMCHANGE 159
#SCROLL 1
#MEASURE 1/4
11,
#BARLINEOFF
#BPMCHANGE 144
#MEASURE 1/4
21,
#BPMCHANGE 126
#MEASURE 1/4
33,

#GOGOEND
#BPMCHANGE 107
#MEASURE 1/4
33,
#BPMCHANGE 60.61
#MEASURE 1/4
00,
#BARLINEON
#MEASURE 5/4

#GOGOSTART
#MEASURE 2/4
#BPMCHANGE 186
20112010,
#MEASURE 4/4
2021202120212021,
2020101020001122,
```
</details>

When this new `.tja` file is passed to `tja2bin`, then the resulting fumen works OK.

Here is the source code for the Python script. I did not write this.

<details>

```python
from abc import abstractmethod
import subprocess
from sys import argv
from os.path import join, exists, basename, split as psplit
from os import mkdir, rename, remove
from shutil import copy2
from shutil import move as rename
from math import gcd

use_2sec = True
use_replace_score = True

if len(argv) < 2:
  print("file path is not present")
  print("usage: py taiko.py songid")
  print("Add another parameter at the end if you don't like 2sec offset")
elif ".tja" not in argv[1]:
  print("wrong file")
  exit(-1)
elif len(argv) == 3:
  use_2sec = False

has_duet = False

forgname = argv[1]
fnewname = (lambda x, y: join(x, '_'+y))(*psplit(forgname))
# TODO: save original .tja and create another one to use
# need to swap the names to have the .bin done correctly

with open(argv[1], 'rb') as f:
  data = f.read().decode("utf-8-sig").replace('\r\n', '\n').replace("#SECTION\n", "")
  if "#START P1" in data:
    has_duet = True
  if use_replace_score:
    data = data.replace("SCOREINIT:\n", "SCOREINIT:500\n").replace("SCOREDIFF:\n", "SCOREDIFF:100\n")


for orig, new_ in ( ("#barlineon", "#barlineon".upper()),
                    ("#barlineoff", "#barlineoff".upper())):
  data = data.replace(orig, new_)

with open(fnewname, 'w', encoding="utf-8") as f:
  f.write(''.join(data))
data = [line 
        for line in data.split('\n') 
          if not any(
            item in line for item in ("MAKER:", "BGMOVIE:", "NOTESDESIGNER:", "#LYRIC")
            )]

if use_2sec:
  data = [line if "OFFSET:" not in line else f"OFFSET:{float(line[7:].strip())-2.0}" for line in data]
  data = [line if "DEMOSTART:" not in line else f"DEMOSTART:{float(line[10:].strip())-2.0}" for line in data]

with open(forgname, 'w', encoding="utf-8") as f:
  f.write('\n'.join(data))

# if not exists(join(".", "debug")):
#   mkdir(join(".", "debug"))

# copy2(argv[1], join("debug", basename(argv[1])))


# Search for weird BPM
class MeasureN:
  m = [4,4]
  tot = 0
  def __init__(self, line) -> None:
    line = line.strip().strip(",")
    self.n = len(line)
    MeasureN.tot += self.n
  @abstractmethod
  def reset():
    MeasureN.m = [4,4]
    MeasureN.tot = 0
  def __repr__(self) -> str:
    nom = MeasureN.m[0]*self.n
    den = MeasureN.tot
    amp = 1
    if nom%den:
      # print("<!> ERROR")
      amp = den // gcd(nom, den)
    return f"#MEASURE {nom*amp//den}/{MeasureN.m[1]*amp}"

data = []
t_data = []
in_chart = False
is_weird = False
with open(argv[1], 'r', encoding="utf-8") as f:
  for line in f:
    line = line.strip()
    if "#START" in line:
      MeasureN.reset()
      in_chart = True
    elif "#END" in line:
      in_chart = False
    elif "#MEASURE" in line:
      MeasureN.m = [int(i) for i in line.strip("#MEASURE ").split('/')]
    if in_chart and "," not in line and "#" not in line and line:
      try:
        int(line)
      except:
        print(f'line was not int but:"{line}"')
        exit(-1)
      is_new = len(t_data) == 0
      t_data.append(MeasureN(line))
      t_data.append(line+',')
      if is_new:
        t_data.append("#BARLINEOFF")
        is_weird = True
    elif "," in line and t_data:
      if line != ",":
        t_data.append(MeasureN(line))
        t_data.append(line)
      data.extend([f"{item}" for item in t_data])
      data.append("#BARLINEON")
      data.append(f"#MEASURE {MeasureN.m[0]}/{MeasureN.m[1]}")
      t_data.clear()
      MeasureN.tot = 0
    elif t_data:
      t_data.append(line)
    else:
      data.append(line)
with open(argv[1], 'w', encoding="utf-8") as f:
  f.write("\n".join(data))


id = argv[1].split('.')[-2].strip("\\")
# if is_weird:
#   with open("_debug.txt", "a+", encoding="utf-8") as file:
#     file.write(id+'\n')
if has_duet:
  in_double = 0
  in_single = False
  in_p1 = False
  in_p2 = False
  solo = ""
  need_p0 = False
  had_p0 = False
  temp_p1 = ""
  with open(argv[1], 'r', encoding="utf-8") as fi, open(join('.', f"{id}_0.tja"), 'w', encoding="utf-8") as fo0, open(join('.', f"{id}_1.tja"), 'w', encoding="utf-8") as fo1, open(join('.', f"{id}_2.tja"), 'w', encoding="utf-8") as fo2:
    for line in fi:
      if "COURSE:" in line:
        print(in_double, in_p1, in_p2, in_single)
        # print(need_p0, had_p0, temp_p1[:20].replace('\n', r'\n'))
        if need_p0 and not had_p0:
          fo0.write(temp_p1)
        temp_p1 = ""
        need_p0 = False
        had_p0 = False
      if "STYLE:Double" in line:
        need_p0 = True
        in_double = 2
      elif "STYLE:Single" in line:
        in_single = True
        had_p0 = True
      elif "#START P1" in line:
        in_p1 = True
        in_p2 = False
        fo1.write("#START\n")
        temp_p1 += "#START\n"
      elif "#START P2" in line:
        in_p2 = True
        in_p1 = False
        fo2.write("#START\n")
      elif in_double:
        if not in_p1:
          fo2.write(line)
        if not in_p2:
          fo1.write(line)
          temp_p1 += line

      elif in_single:
        fo0.write(line)
      else:
        fo1.write(line)
        fo2.write(line)
        fo0.write(line)

      if "#END" in line:
        if in_double:
          in_double -= 1
        in_p1, in_p2 = False, False
        in_single = False

    if need_p0 and not had_p0:
      fo0.write(temp_p1)

# exit(0)
if not exists("solo"):
  mkdir("solo")
if not exists("duet"):
  mkdir("duet")
if not has_duet:
  subprocess.call([join(".", "taiko.exe"), "--be", argv[1]], shell=False)
else:
  for fname in (f"{id}_0.tja", f"{id}_1.tja", f"{id}_2.tja"):
    subprocess.call([join(".", "taiko.exe"), "--be", join(".", fname)], shell=False)
    remove(join(".", fname))


for diff in "enhmx"[::-1]:

  nodiff = f"{id}.bin"
  srcname = f"{id}_{diff}.bin"

  dstname = f"{id}_{diff}"
  if diff == "x":
    dstname = f"ex_{id}_m"
  # if diff == "m":
  #   srcname = f"{id}.bin"

  if exists(join(".", nodiff)) and diff == 'm':
    print("only one diff")
    copy2(join(".", nodiff), join(".", "duet", f"{dstname}_1.bin"))
    copy2(join(".", nodiff), join(".", "duet", f"{dstname}_2.bin"))
    rename(nodiff, join("solo", f"{dstname}.bin"))
  elif not has_duet:
    if not exists(join(".", srcname)):
      print("difficulty", diff, "not found")
      continue
    copy2(join(".", srcname), join(".", "duet", f"{dstname}_1.bin"))
    copy2(join(".", srcname), join(".", "duet", f"{dstname}_2.bin"))
    rename(srcname, join("solo", f"{dstname}.bin"))
  else:
    # if not exists(join(".", f"{id}_0_{diff}.bin")):
    #   print("difficulty", diff, "not found")
    #   continue
    if exists(join(".", f"{id}_0_{diff}.bin")):
      rename(join(".", f"{id}_0_{diff}.bin"), join(".", "solo", f"{dstname}.bin"))
    else:
      print("difficulty", diff, "(0) not found")
    if exists(join(".", f"{id}_1_{diff}.bin")):
      rename(join(".", f"{id}_1_{diff}.bin"), join(".", "duet", f"{dstname}_1.bin"))
    else:
      print("difficulty", diff, "(1) not found")
    if exists(join(".", f"{id}_2_{diff}.bin")):
      rename(join(".", f"{id}_2_{diff}.bin"), join(".", "duet", f"{dstname}_2.bin"))
    else:
      print("difficulty", diff, "(2) not found")

rename(fnewname, forgname)
```
</details>

To fix this in TakoTako, I guess we would have to rewrite the Python script in C#?

**Comment by vivaria** [comment]
Hi there! This issue is actually 2 issues in one:

1. **200ms offset**: ~~I believe this is due to TakoTako's processing of the `.ogg` file.~~ (Incorrect, see [comment](https://github.com/Fluto/TakoTako/issues/16#issuecomment-1636960820) below.) 
2. **BPM desync near the end**: I took a few months to reverse engineer the fumen format, and I discovered that there is a [specific adjustment](https://github.com/vivaria/tja2fumen/blob/ee265327deed3ceb3e9623fdbbf18d0d03f6d7d0/src/tja2fumen/converters.py#L161-L182) that needs to be done whenever there is a #BPMCHANGE. I have tested this behavior extensively, and I created a new converter that doesn't have any of the #BPMCHANGE issues that the current `tja2bin.exe` converter has. Please try it out! https://github.com/vivaria/tja2fumen

So, we're halfway there, and I'm hoping to work together with @Fluto to fix the first issue, too. 😄

**Comment by vivaria** [comment]
I've discovered the cause of the full-song desync. tl;dr: 

- Each TJA's OFFSET value (multiplied by 10) is being added to the `fumenOffsetPos` field in the `data.json` metadata file.
- Songs with a larger OFFSET value experience greater desync.
- Rokuchounen has a large OFFSET of -11.4s, so this results in a desync of 114ms, which is very noticeable. 
- Setting `fumenOffsetPos` to 0 fixes the desync issues entirely.

Here is a more in-depth write-up:

<details>
<summary> Investigation </summary>

----

In talking to @Fluto in Discord DMs, he identified this line where `millisecondsAddedSilence` is set:

<details>

https://github.com/Fluto/TakoTako/blob/c269bcab60530877a16c2a473c84796b94d0a5ce/TJAConvert/Program.cs#L120

https://github.com/Fluto/TakoTako/blob/c269bcab60530877a16c2a473c84796b94d0a5ce/TJAConvert/Program.cs#L26-L27

* If `OFFSET` is greater than `-1`, then 2000ms of silence will be added
* If `OFFSET` is less than `-1`, then 0ms will be added

In Rokuchounen's case, the OFFSET is `-11.4302`, which is less than `-1`, so 0ms is added to the song (2:22 total song length). However, if I manually edit the line to add 2000ms to the song (2:24 total song length), then the offset is naturally _much much worse_! (e.g. ~200ms off before, ~2200ms off after.) 

</details>

Given these details, I don't think `millisecondsAddedSilence` is the cause of the desync. 

----

Separately, I *did* notice a specific field in the `data.json` metadata file called `fumenOffsetPos`:

<details>

- In my official fumen, `fumenOffsetPos` has a value of 0
- In the converted TJA, `fumenOffsetPos` has a value of -114

The different (114ms) seems awfully close to the ~200ms offset that was noticed. So, I tried updating the converted TJA's `data.json` to use a value of 0, and... it's synced up perfectly!!!

</details>

Aha! `fumenOffsetPos` is definitely the culprit here.

----

Just to confirm that `millsecondsAddedSilence` isn't causing any desync, I tried converting a song with OFFSET > `-1` that I know is synced up decently well (`Arc.tja`). Its offset is `-0.649`, so, as described previously, an extra 2000ms of silence is added to the audio file. And, accordingly, Arc's fumenOffsetPos value is `1994`. This means that `millsecondsAddedSilence` is compensated for, and isn't the cause of the problem. 

(In fact, I think the added silence is just to make sure that song with small offsets don't start too early, which would disorient the player. This is a good feature, @Fluto!)

---

However, given the behavior with both Arc and Rokuchounen, I think what is happening is that the `fumenOffsetPos` value is computed by adding `millisecondsAddedSilence` + `OFFSET` (but... off by a factor of 10).

- Rokuchounen: 0 + -11.4302s =  -114
- Arc: 2000 + -0.649s = 1994

My theory is that the larger the OFFSET, the more desync is noticeable. This is why Rokuchounen (-11.4s) and Doppelgangers (-8.7s) result in a very noticeable desync of 114ms and 87ms respectively, but a song like Arc (-0.6s) results in a desync of 6ms, which is hardly noticeable.

Given that the Ultramix songs use a fixed offset value of 0, I think the solution here is to _not_ add the OFFSET, and instead use a fixed value of 0, too. (Or, a value of 2000 if the silence is added.) 

I've tried using a value of 0 on other songs, too, and it fixes the offset entirely!!

</details>

**Comment by Fluto** [comment]
Issues addressed in f728d694b38da732d06a687b14ebbe676a6ff435. This fix will be available in the next update.
Thank you everyone for their feedback! And thank you @vivaria for working on a fix.

**Comment by vivaria** [comment]
For anyone using an older version of TakoTako (v3.2.0 and below), I wrote a quick little script to fix the values of the `fumenOffsetPos` (see https://github.com/Fluto/TakoTako/issues/16#issuecomment-1636960820). 

This way, all of your converted TJAs will be in sync, even when using an older version of TakoTako: https://gist.github.com/vivaria/663ac4ffaf37345f0abd55e1cc8dfe32
